content/research/market-health/posts/2026-05-11-okx/analysis.py

```
@@ -0,0 +1,114 @@
+        for c in batch:
+            candles.append({"ts": c[0], "open": float(c[1]), "high": float(c[2]),
+                            "low": float(c[3]), "close": float(c[4]), "vol": float(c[5])})
+        end_time = batch[0][0] - 1
+    return sorted(candles, key=lambda x: x["ts"])[:limit]
+
+
+def load_or_fetch() -> dict:
+    if CACHE.exists():
+        print(f"Loading cached data from {CACHE.name}")
+        with open(CACHE, "rb") as f:
+            return pickle.load(f)
```

Ruff (0.15.12)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)